### PR TITLE
catalog-node: update catalogAnalysisExtensionPoint

### LINKED ISSUE #24505: 🐛 Bug Report:  CatalogAnalysisExtensionPoint accepts only ScmLocationAnalyzer
### 📜 Description

The old [setLocationAnalyzers](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/service/CatalogBuilder.ts#L260) accepts any _LocationAnalyzer_ instance.

While the new CatalogAnalysisExtensionPoint and the [addLocationAnalyzer](https://github.com/backstage/backstage/blob/master/plugins/catalog-node/src/extensions.ts#L82) accepts only _ScmLocationAnalyzer_. 

The PR should:
-  Update the interface to accept any _LocationAnalyzer_
-  Add a separate method for scm analyzer specifically

### 👍 Expected behavior

CatalogAnalysisExtensionPoint should accept any LocationAnalyzer implementation

### 👎 Actual Behavior with Screenshots

CatalogAnalysisExtensionPoint accepts only ScmAnalyzers

### 👟 Reproduction steps

index.ts
```ts
import { createBackend } from '@backstage/backend-defaults';

// Create your backend instance
const backend = createBackend();

// Install all desired features
backend.add(import('@backstage/plugin-app-backend/alpha'));
backend.add(import('@backstage/plugin-auth-backend'));
backend.add(import('@backstage/plugin-catalog-backend-module-scaffolder-entity-model'));
backend.add(import('./plugins/catalog'));
backend.add(import('@backstage/plugin-catalog-backend/alpha'));

backend.add(import('@backstage/plugin-proxy-backend/alpha'));
backend.add(import('@backstage/plugin-scaffolder-backend/alpha'));
backend.add(import('@backstage/plugin-search-backend-module-catalog/alpha'));
backend.add(import('@backstage/plugin-search-backend/alpha'));
backend.add(import('@backstage/plugin-techdocs-backend/alpha'));

// Start up the backend
backend.start();
```

plugins/catalog.ts
```ts
import { ScmIntegrations } from '@backstage/integration';
import {CustomRepoLocationAnalyzer} from '../modules/CustomRepoLocationAnalyzer';
import { GithubLocationAnalyzer } from '@backstage/plugin-catalog-backend-module-github';
import { createBackendModule } from '@backstage/backend-plugin-api';
import { catalogAnalysisExtensionPoint } from '@backstage/plugin-catalog-node/alpha';
import {
    coreServices,
} from '@backstage/backend-plugin-api';

export default createBackendModule({
    pluginId: 'catalog',
    moduleId: 'example-custom-analyzer',
    register(env) {
        env.registerInit({
            deps: {
                catalog: catalogAnalysisExtensionPoint,
                config: coreServices.rootConfig,
                logger: coreServices.logger,
                tokenManager: coreServices.tokenManager,
                discovery: coreServices.discovery,
            },
            async init({  config, logger, tokenManager, discovery }) {
                const analyzer = new CustomRepoLocationAnalyzer(
                    logger,
                    ScmIntegrations.fromConfig(config),
                    [
                        new GithubLocationAnalyzer({
                            discovery: discovery,
                            tokenManager: tokenManager,
                            config: config,
                        }),
                    ],
                );
                console.log(analyzer);
                // Blocker: This does not work as addLocationAnalyzer requires ScmAnalyzer
                // catalog.addLocationAnalyzer(analyzer);
            },
        });
    },
});
```

CustomRepoLocationAnalyzer.ts
```ts
import parseGitUrl from 'git-url-parse';
import { Entity } from '@backstage/catalog-model';
import { ScmIntegrationRegistry } from '@backstage/integration';
import {
  AnalyzeLocationRequest,
  AnalyzeLocationResponse,
} from '@backstage/plugin-catalog-common';
import { ScmLocationAnalyzer } from '@backstage/plugin-catalog-node';
import { LoggerService } from '@backstage/backend-plugin-api';
import { LocationAnalyzer } from '@backstage/plugin-catalog-backend';

export class CustomRepoLocationAnalyzer implements LocationAnalyzer {
  private readonly logger: LoggerService;
  private readonly scmIntegrations: ScmIntegrationRegistry;
  private readonly analyzers: ScmLocationAnalyzer[];

  constructor(
    logger: LoggerService,
    scmIntegrations: ScmIntegrationRegistry,
    analyzers: ScmLocationAnalyzer[],
  ) {
    this.logger = logger;
    this.scmIntegrations = scmIntegrations;
    this.analyzers = analyzers;
  }
  async analyzeLocation(
    request: AnalyzeLocationRequest,
  ): Promise<AnalyzeLocationResponse> {
    const integration = this.scmIntegrations.byUrl(request.location.target);
    const { owner, name } = parseGitUrl(request.location.target);

    let annotationPrefix;
    switch (integration?.type) {
      case 'azure':
        annotationPrefix = 'dev.azure.com';
        break;
      case 'bitbucket':
        annotationPrefix = 'bitbucket.org';
        break;
      case 'github':
        annotationPrefix = 'github.com';
        break;
      case 'gitlab':
        annotationPrefix = 'gitlab.com';
        break;
      default:
        break;
    }

    const analyzer = this.analyzers.find(a =>
      a.supports(request.location.target),
    );
    if (analyzer) {
      const analyzerResult = await analyzer.analyze({
        url: request.location.target,
      });
      if (analyzerResult.existing.length > 0) {
        this.logger.debug(
          `entity for ${request.location.target} already exists.`,
        );
        return {
          existingEntityFiles: analyzerResult.existing,
          generateEntities: [],
        };
      }
    }

    const entity: Entity = {
      apiVersion: 'backstage.io/v1alpha1',
      kind: 'Component',
      metadata: {
        name: name,
      },
      spec: { type: 'other', lifecycle: 'emerging' }, // This is where we modify the default file contents of catalog-info.yaml
    };

    if (annotationPrefix) {
      entity.metadata.annotations = {
        [`${annotationPrefix}/project-slug`]: `${owner}/${name}`,
      };
    }

    this.logger.debug(`entity created for ${request.location.target}`);
    return {
      existingEntityFiles: [],
      generateEntities: [{ entity, fields: [] }],
    };
  }
}
```

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Fixes #24505

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-node/api-report-alpha.md [hunk 1]
```diff
@@ -20,8 +21,8 @@ import { Validators } from '@backstage/catalog-model';
 
 // @alpha (undocumented)
 export interface CatalogAnalysisExtensionPoint {
-  // (undocumented)
-  addLocationAnalyzer(analyzer: ScmLocationAnalyzer): void;
+  addScmLocationAnalyzer(analyzer: ScmLocationAnalyzer): void;
```
Comment (freben) [inline plugins/catalog-node/api-report-alpha.md:24]: I was hoping to not have this at all, but instead opting for

```ts
setLocationAnalyzer(
  DefaultLocationAnalyzer.create({
    deps,
    analyzers: [new CustomThing()],
 })
);
```

Or similar and heading toward that as a general pattern (eg for the catalog model etc)

This odd duality of "add child - or replace parent and then children are ignored" would be nice to get rid of.

Comment (Rugvip) [reply plugins/catalog-node/api-report-alpha.md:24]: This lets builders ship support for analysis of individual providers that can be installed without manual module wiring. We could ship the default one with a GitHub analyzer from the GitHub module but then you can't combine support form multiple SCMs without code changes.

Comment (Rugvip) [reply plugins/catalog-node/api-report-alpha.md:24]: How do you mean?

Comment (freben) [reply plugins/catalog-node/api-report-alpha.md:24]: `setLocationAnalyzer((scmAnalyzers: ScmLocationAnalyzer[]) => Location Analyzer): void` so to speak.

Maybe I'm misunderstanding the interfaces somewhere and should wait until I get a chance to check the branch out locally and poke at it :)

Comment (Rugvip) [reply plugins/catalog-node/api-report-alpha.md:24]: Ah yep I guess we could do that. Ofc there's no guarantee that the `LocationAnalyzer` implementation can or will use the SCM analyzers, but feels like an improvement still.

Comment (Rugvip) [reply plugins/catalog-node/api-report-alpha.md:24]: Mkay updated with a new API for `setLocationAnalyzer`

**Review by freben — APPROVED** [review]

**Pull request merged**